ush/ioda/bufr2ioda/bufr2ioda_conventional_prepbufr_ps.py

```diff
@@ -0,0 +1,792 @@
+
+        elif subsets[i] == "SFCSHP":
+            logger.debug("Making QuerySet for SFCSHP")
+
+            # ObsType
+            r.add('observationType', '*/TYP')
+
+            # MetaData
+            r.add('stationIdentification', '*/SID')
+            r.add('prepbufrDataLevelCategory', '*/CAT')
+            r.add('temperatureEventCode', '*/T___INFO/T__EVENT{1}/TPC')
+            r.add('latitude', '*/YOB')
+            r.add('longitude', '*/XOB')
+            r.add('obsTimeMinusCycleTime', '*/DHR')
+            r.add('heightOfStation', '*/Z___INFO/Z__EVENT{1}/ZOB')
+            r.add('pressure', '*/P___INFO/P__EVENT{1}/POB')
+
+            # QualityMarker
+            r.add('qualityMarkerStationPressure', '*/P___INFO/P__EVENT{1}/PQM')
+            r.add('qualityMarkerAirTemperature', '*/T___INFO/T__EVENT{1}/TQM')
+            r.add('qualityMarkerVirtualTemperature', '*/T___INFO/T__EVENT{1}/TQM')
+
+            # ObsValue
+            r.add('stationElevation', '*/ELV')
+            r.add('stationPressure', '*/P___INFO/P__EVENT{1}/POB')
+            r.add('airTemperature', '*/T___INFO/T__EVENT{1}/TOB')
+
+        elif subsets[i] == "ADPUPA":
+            logger.debug("Making QuerySet for ADPUPA")
+            # ObsType
+            s.add('observationType', '*/TYP')
+
+            # MetaData
+            s.add('stationIdentification', 'ADPUPA/SID')
+            s.add('prepbufrDataLevelCategory', '*/PRSLEVEL/CAT')
+            s.add('temperatureEventCode', '*/PRSLEVEL/T___INFO/T__EVENT{1}/TPC')
+            s.add('latitude', '*/PRSLEVEL/DRFTINFO/YDR')
+            s.add('longitude', '*/PRSLEVEL/DRFTINFO/XDR')
+            s.add('heightOfStation', '*/PRSLEVEL/Z___INFO/Z__EVENT{1}/ZOB')
+            s.add('timeOffset', '*/PRSLEVEL/DRFTINFO/HRDR')
+            s.add('releaseTime', '*/PRSLEVEL/DRFTINFO/HRDR')
+            s.add('pressure', '*/PRSLEVEL/P___INFO/P__EVENT{1}/POB')
+
+            # QualityMarker
+            s.add('qualityMarkerStationPressure', '*/PRSLEVEL/P___INFO/P__EVENT{1}/PQM')
+            s.add('qualityMarkerAirTemperature', '*/PRSLEVEL/T___INFO/T__EVENT{1}/TQM')
+            s.add('qualityMarkerVirtualTemperature', '*/PRSLEVEL/T___INFO/T__EVENT{1}/TQM')
+
+            # ObsValue
+            s.add('stationElevation', '*/ELV')
+            s.add('stationPressure', '*/PRSLEVEL/P___INFO/P__EVENT{1}/POB')
+            s.add('airTemperature', '*/PRSLEVEL/T___INFO/T__EVENT{1}/TOB')
+
+    end_time = time.time()
+    running_time = end_time - start_time
+    logger.debug(f"Running time for making QuerySet: {running_time} seconds")
+
+    # ==============================================================
+    # Open the BUFR file and execute the QuerySet to get ResultSet
+    # Use the ResultSet returned to get numpy arrays of the data
+    # ==============================================================
+    start_time = time.time()
+
+    logger.debug(f"Executing QuerySet to get ResultSet ...")
+    with bufr.File(DATA_PATH) as f:
```

Comment:
Any reason why these need opened up 3x? Or is this a BUFR quirk?

Reply:
It's a quirk. From what I understand, it goes through the whole thing on the ADPSFC data query, and then at the end it has to rewind. I found it easier to open it again.